### PR TITLE
mon: include device class in tree view; hide shadow hierarchy

### PR DESCRIPTION
**Pull request by liewegas**
- Do not show each per-class variation of the hieararchy
- Do include a CLASS column in the tree view

### CODE REVIEW

**Comment by liewegas** [comment]
retest this please

**Comment by liewegas** [comment]
retest this please

**Comment by liewegas** [comment]
retest this please

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-07-09_19:14:46-rados-wip-sage-testing-distro-basic-smithi/

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-07-09_19:14:46-rados-wip-sage-testing-distro-basic-smithi/ is ok but manual test failed to fix the actual bug

File: src/test/cli/osdmaptool/tree.t [hunk 1]
```diff
@@ -4,16 +4,16 @@
 
   $ osdmaptool --tree=plain om
   osdmaptool: osdmap file 'om'
-  ID WEIGHT  TYPE NAME              UP/DOWN REWEIGHT PRIMARY-AFFINITY 
-  -1 3.00000 root default
```
Comment (tchaikov) [inline src/test/cli/osdmaptool/tree.t]:
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/cli/osdmaptool/tree.t: failed
--- /home/jenkins-build/build/workspace/ceph-pull-requests/src/test/cli/osdmaptool/tree.t
+++ /home/jenkins-build/build/workspace/ceph-pull-requests/src/test/cli/osdmaptool/tree.t.err
@@ -4,13 +4,13 @@
 
   $ osdmaptool --tree=plain om
   osdmaptool: osdmap file 'om'
-  ID CLASS WEIGHT  TYPE NAME              UP/DOWN REWEIGHT PRI-AFF
-  -1       3.00000 root default
-  -3       3.00000     rack localrack
-  -2       3.00000         host localhost
-   0       1.00000             osd.0          DNE        0
-   1       1.00000             osd.1          DNE        0
-   2       1.00000             osd.2          DNE        0
+  ID CLASS WEIGHT  TYPE NAME              UP/DOWN REWEIGHT PRI-AFF 
+  -1       3.00000 root default                                    
+  -3       3.00000     rack localrack                              
+  -2       3.00000         host localhost                          
+   0       1.00000             osd.0          DNE        0         
+   1       1.00000             osd.1          DNE        0         
+   2       1.00000             osd.2          DNE        0         
 
   $ osdmaptool --tree=json om
   osdmaptool: osdmap file 'om'
```

i don't think we can remove the trailing spaces here.

File: src/crush/CrushTreeDumper.h [hunk 2]
```diff
@@ -150,6 +150,10 @@ namespace CrushTreeDumper {
   inline void dump_item_fields(const CrushWrapper *crush,
 			       const Item &qi, Formatter *f) {
     f->dump_int("id", qi.id);
+    const char *c = crush->get_item_class(qi.id);
+    if (!c)
+      c = "";
+    f->dump_string("device_class", c);
```
Comment (tchaikov) [inline src/crush/CrushTreeDumper.h]:
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/mon/osd-crush.sh:229: TEST_crush_tree:  ceph osd crush tree --format=xml
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/mon/osd-crush.sh:230: TEST_crush_tree:  xmlstarlet val -e -r /home/jenkins-build/build/workspace/ceph-pull-requests/src/test/mon/osd-crush-tree.rng -
-:1.166: Did not expect element device_class there
-:1.166: Did not expect element name there
-:1.166: Did not expect text in element  TODO  content
-:1.166: Did not expect element type there
-:1.166: Did not expect text in element  TODO  content
-:1.166: Did not expect element type_id there
-:1.166: Did not expect text in element  TODO  content
-:1.166: Did not expect element items there
- - invalid
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/mon/osd-crush.sh:230: TEST_crush_tree:  return 1
```

we also need to update the rng file accordingly

```diff
diff --git a/src/test/mon/osd-crush-tree.rng b/src/test/mon/osd-crush-tree.rng
index d050ed2052..14ea5f35a0 100644
--- a/src/test/mon/osd-crush-tree.rng
+++ b/src/test/mon/osd-crush-tree.rng
@@ -14,6 +14,9 @@
       <element name="id">
         <text/>
       </element>
+      <element name="device_class">
+        <text/>
+      </element>
       <element name="name">
         <text/>
       </element>
@@ -36,6 +39,9 @@
       <element name="id">
         <text/>
       </element>
+      <element name="device_class">
+        <text/>
+      </element>
       <element name="name">
         <text/>
       </element>
```

**Comment by tchaikov** [comment]
one of the test failures in http://pulpito.ceph.com/sage-2017-07-09_19:14:46-rados-wip-sage-testing-distro-basic-smithi/ is addressed by https://github.com/ceph/ceph/pull/16235

**Comment by tchaikov** [comment]
once the "make check" failures are fixed, lgtm.

**Pull request merged**